### ISSUE TITLE
create helper functions to move between domain size

### ISSUE THREAD

**Issue by mimoo** [o1-labs/proof-systems#886]
In https://github.com/o1-labs/proof-systems/blob/master/kimchi/src/circuits/domains.rs#L1 we define 4 domains of different sizes: d1, d2, d4, and d8. Each domain having twice as many elements are the previous one.

Evaluations in a specific domain size can quickly be casted in a _smaller_ domain by taking all the i-th elements, where i is the difference of size between the two domains. For example to go from d4 to d1, you'd pick every 4th element, and to go from d8 to d1 you'd pick every 8th element.

This results in a _lot_ of code that looks like this and is hard to parse:

```rust
let s = (plnm.domain().size / domain.size) as usize;
let v: Vec<_> = (0..(domain.size())).map(|i| plnm.evals[s * i]).collect();
```

sometimes it is even interlaced with some scaling:

```rust
x + evals.evals[(scale * i + (domain as usize) * shift) % evals.evals.len()]
```

it would be good to create helper functions to hide the low-level manipulation and make the code around it easier to parse. I can try to find more example of this pattern being used in the codebase, but grepping for these would be a good starting point!

**Comment by duguorong009** [comment]
@mimoo 
I would like to take this issue!

**Comment by duguorong009** [comment]
@mimoo
Atm, I found that the domain move is mostly done in `[commitment.rs](https://github.com/o1-labs/proof-systems/blob/master/poly-commitment/src/commitment.rs)` and `[expr.rs](https://github.com/o1-labs/proof-systems/blob/master/kimchi/src/circuits/expr.rs)`.
I think these appearances are mainly attributed to that the domain move/conversion is needed for polynomial commitment & aggregation(shrinking the polynomial degree) processes.
It would be best if you can confirm above findings & indicate me more appearances anywhere.

**Comment by duguorong009** [comment]
@mimoo 
I tried to create a helper function for the issue.
TBH, I think this is a bit of `overkill`.
The reason is that the most of these patterns are used in kind of `shifting` elements.
```
let scale = (domain as usize) / (res_domain.0 as usize);
assert!(scale != 0);
let v: Vec<_> = (0..n)
  .into_par_iter()
  .map(|i| {
     x + evals.evals[(scale * i + (domain as usize) * shift) % evals.evals.len()]
  })
.collect();
```
Please let me know if I am wrong.

**Comment by mimoo** [comment]
the helper function could just take a scalar for shifting if needed no? The reason why I don't think it's overkill is that it's tedious to parse these lines of code every time AND sometimes people forget to optimize such code with `par_iter`

I think the locations you found are a good starting point, I can find more after : o

**Comment by duguorong009** [comment]
Thanks @mimoo 
I will try to create create a helper.

**Comment by duguorong009** [comment]
@mimoo 
Please check the PR #899 I created for the issue.

**Comment by duguorong009** [comment]
@mimoo 
TBH, I think this helper function is a bit inappropriate.
The reason is that the helper function creates the extra `Vec`s & add more loops.

For example: 
[Here](https://github.com/o1-labs/proof-systems/blob/master/kimchi/src/circuits/expr.rs#L1008-L1020), the code is just creating the final result within **one** loop and **no extra vec**.
```
let scale1 = (d1 as usize) / (res_domain.0 as usize);
assert!(scale1 != 0);
let scale2 = (d2 as usize) / (res_domain.0 as usize);
assert!(scale2 != 0);

let n = res_domain.1.size();
let v: Vec<_> = (0..n)
    .into_par_iter()
    .map(|i| {
        es1.evals[(scale1 * i + (d1 as usize) * s1) % es1.evals.len()]
            + es2.evals[(scale2 * i + (d2 as usize) * s2) % es2.evals.len()]
    })
    .collect();
```

But, in the [PR](https://github.com/duguorong009/proof-systems/blob/create-move-domain-size-helper/kimchi/src/circuits/expr.rs#L1015-L1032), the code creates **2 extra vecs** & runs **3 loops**.
```
  let v1 = evals_domain_size_cast(
    &es1.evals,
    d1 as usize,
    res_domain.0 as usize,
    res_domain.1.size(),
    s1,
);

let v2 = evals_domain_size_cast(
    &es2.evals,
    d2 as usize,
    res_domain.0 as usize,
    res_domain.1.size(),
    s2,
);

let n = res_domain.1.size();
let v: Vec<_> = (0..n).into_par_iter().map(|i| v1[i] + v2[i]).collect();
```

I worry if it would not affect the runtime/memory of the proving system, & efficiency of system.(maybe slower?)

Please let me know if I am wrong.

**Comment by mimoo** [comment]
more occurences in that PR: https://github.com/o1-labs/proof-systems/pull/944/files

**Comment by mimoo** [comment]
@duguorong009 I think I missed a number of notifications : o I'm just seeing this now sorry!

**Comment by mimoo** [comment]
proposed this interface: https://github.com/o1-labs/proof-systems/pull/964

**Comment by jspada** [comment]
> proposed this interface: #964

To help get this over the line, I've created PR #965 where I've branched from PR #964  and fully integrated it into the expression framework.  There were quite a few different use cases and scenarios in there, which necessitated some extensions to the proposed interface.

Hope this helps!

**Comment by duguorong009** [comment]
@mimoo @jspada 
Thanks for the valuable guides & PRs!

I looked into the code & it looks great to me.
Please go for merge if you are confident about the PR #965 .
(I think PR #964 could also be closed thanks to PR #965)

Thanks again! :+1:
